File: python/pages/login_page.py
from playwright.sync_api import Page, Locator


class LoginPage:
    def __init__(self, page: Page):
        self._page = page
        self._username_input: Locator = page.get_by_role('textbox', name='Username')
        self._password_input: Locator = page.get_by_role('textbox', name='Password')
        self._login_button:   Locator = page.get_by_role('button',  name='Login')

    def goto(self):
        self._page.goto('/web/index.php/auth/login')

    def login(self, username: str, password: str):
        self._username_input.fill(username)
        self._password_input.fill(password)
        self._login_button.click()

File: csharp/Pages/LoginPage.cs
using Microsoft.Playwright;

namespace PlaywrightComparison.Pages;

public class LoginPage
{
    private readonly IPage _page;
    private readonly ILocator _usernameInput;
    private readonly ILocator _passwordInput;
    private readonly ILocator _loginButton;

    public LoginPage(IPage page)
    {
        _page          = page;
        _usernameInput = page.GetByRole(AriaRole.Textbox, new() { Name = "Username" });
        _passwordInput = page.GetByRole(AriaRole.Textbox, new() { Name = "Password" });
        _loginButton   = page.GetByRole(AriaRole.Button,  new() { Name = "Login" });
    }

    public async Task GotoAsync() =>
        await _page.GotoAsync("/web/index.php/auth/login");

    public async Task LoginAsync(string username, string password)
    {
        await _usernameInput.FillAsync(username);
        await _passwordInput.FillAsync(password);
        await _loginButton.ClickAsync();
    }
}
